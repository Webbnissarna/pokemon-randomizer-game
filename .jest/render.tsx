import { render as rtlRender } from "@testing-library/react";
import { InitializeColorMode, ThemeProvider } from "theme-ui";

import theme from "../utils/theme";

import "./style.css";

export function render(ui: React.ReactElement) {
  return rtlRender(
    <>
      <InitializeColorMode />
      <ThemeProvider theme={theme}>{ui}</ThemeProvider>
    </>
  );
}
