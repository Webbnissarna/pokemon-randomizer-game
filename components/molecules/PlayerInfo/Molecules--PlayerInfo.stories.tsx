import { Flex } from "theme-ui";
import PlayerInfo from ".";
import { mockPokemon } from "../../../utils/testHelper";
import type { Variant } from "./PlayerInfo.types";

interface ControlProps {
  playerName: string;
  variant: Variant;
  ballCount: number;
  pickedCount: number;
}

export function Default({
  playerName,
  variant,
  ballCount,
  pickedCount,
}: ControlProps): JSX.Element {
  return (
    <Flex sx={{ flexDirection: "column", gap: 20 }}>
      <h1>Custom</h1>
      <PlayerInfo
        variant={variant}
        ballCount={ballCount}
        playerName={playerName}
        pickedPokemon={
          pickedCount > 0 ? Array(pickedCount).fill(mockPokemon) : []
        }
      />

      <h1>Preset</h1>
      <PlayerInfo
        variant="tl"
        ballCount={6}
        playerName={"Player 1 Name"}
        pickedPokemon={[mockPokemon, mockPokemon, mockPokemon]}
      />
      <PlayerInfo
        variant="br"
        ballCount={6}
        playerName={"Player 2 Name"}
        pickedPokemon={[mockPokemon, mockPokemon]}
      />
    </Flex>
  );
}

Default.args = {
  playerName: "Player Name",
  variant: "tl",
  ballCount: 6,
  pickedCount: 3,
} as ControlProps;

Default.argTypes = {
  variant: {
    options: ["tl", "br"],
    control: { type: "select" },
  },
};
