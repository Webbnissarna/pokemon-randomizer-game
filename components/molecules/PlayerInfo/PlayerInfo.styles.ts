import type { ThemeUIStyleObject } from "theme-ui";
import type { Variant } from "./PlayerInfo.types";

export const variants: Record<Variant, ThemeUIStyleObject> = {
  tl: {
    flexDirection: "column",
    textAlign: "left",
  },
  br: {
    flexDirection: "column-reverse",
    textAlign: "right",
  },
};

export const containerStyle: ThemeUIStyleObject = {
  width: "fit-content",
  gap: "md",
};

export const textStyle: ThemeUIStyleObject = {
  fontSize: "lg",
};

export const ballContainerStyle: ThemeUIStyleObject = { gap: "md" };
