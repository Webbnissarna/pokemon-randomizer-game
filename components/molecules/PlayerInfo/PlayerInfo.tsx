import { Flex, Text } from "theme-ui";
import Ball from "../../atoms/Ball";
import {
  ballContainerStyle,
  containerStyle,
  textStyle,
  variants,
} from "./PlayerInfo.styles";
import type { PlayerInfoProps } from "./PlayerInfo.types";

export default function PlayerInfo({
  playerName,
  variant,
  ballCount,
  pickedPokemon,
}: PlayerInfoProps): JSX.Element {
  const unpickedCount = ballCount - pickedPokemon.length;

  return (
    <Flex sx={{ ...variants[variant], ...containerStyle }}>
      <Text sx={textStyle}>{playerName}</Text>
      <Flex sx={ballContainerStyle}>
        {pickedPokemon.map((_, i) => (
          <Ball key={i} size="sm" disabled={false} onClick={() => undefined} />
        ))}

        {unpickedCount > 0 &&
          Array(unpickedCount)
            .fill(0)
            .map((_, i) => (
              <Ball
                key={i}
                size="sm"
                disabled={true}
                onClick={() => undefined}
              />
            ))}
      </Flex>
    </Flex>
  );
}
