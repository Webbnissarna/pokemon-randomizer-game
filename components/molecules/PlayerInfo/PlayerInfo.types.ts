import type { GeneratedPokemon } from "../../../utils/Generator/generator.types";

export type Variant = "tl" | "br";

export interface PlayerInfoProps {
  playerName: string;
  variant: Variant;
  ballCount: number;
  pickedPokemon: GeneratedPokemon[];
}
