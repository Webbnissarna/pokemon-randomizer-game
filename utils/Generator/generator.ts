import type {
  Ability,
  BasicMoveInfo,
  BasicPokemonInfo,
  MoveInfo,
  Nature,
  PokemonInfo,
  RNG,
} from "./pokemon.types";
import type {
  GeneratedPokemon,
  GeneratedPokemonNoExport,
  MoveListResolver,
  BasicMoveResolver,
  PokemonResolver,
  MoveResolver,
} from "./generator.types";
import {
  getUniquesFromLists,
  makeRandomUniqueList,
  getRandomFromList,
  isValid,
} from "./helper";

export function getShowdownExport({
  info,
  ability,
  nature,
  moves,
}: GeneratedPokemonNoExport) {
  return `${info.name}
Ability: ${ability.name.en}
EVs: 1 HP / 1 Atk / 1 Def / 1 SpA / 1 SpD / 1 Spe
${nature.name.en} Nature
${moves.map((move) => `- ${move.name.en}`).join("\n")}`;
}

async function getPossibleMoves(
  info: PokemonInfo,
  moveResolver: BasicMoveResolver,
  moveListResolver: MoveListResolver
): Promise<BasicMoveInfo[]> {
  const regularMoveNames = getUniquesFromLists(
    info.moves.leveling.map((m) => m.move),
    info.moves.tm.map((m) => m.move)
  );

  const regularMoveList = await Promise.all(regularMoveNames.map(moveResolver));

  const typedMoveList = (
    await Promise.all(info.types.map(moveListResolver))
  ).flat();

  const uniqueMovesSet = new Set([...regularMoveList, ...typedMoveList]);
  const allPossibleMovesList = [...uniqueMovesSet];
  return allPossibleMovesList;
}

async function getRandomMoves(
  random: RNG,
  moveList: BasicMoveInfo[],
  count: number,
  moveResolver: MoveResolver
): Promise<MoveInfo[]> {
  const randomMoveIndices = makeRandomUniqueList(
    random,
    0,
    moveList.length - 1,
    count
  );

  const moves = randomMoveIndices.map((moveIndex) => moveList[moveIndex]);
  const moveNames = moves.map((move) => move.name.en).filter<string>(isValid);
  const fullMoves = await Promise.all(
    moveNames.map((moveName) => moveResolver(moveName))
  );

  return fullMoves;
}

export async function generatePokemon(
  random: RNG,
  possiblePokemon: BasicPokemonInfo[],
  possibleNatures: Nature[],
  possibleAbilities: Ability[],
  pokemonResolver: PokemonResolver,
  basicMoveResolver: BasicMoveResolver,
  moveResolver: MoveResolver,
  moveListResolver: MoveListResolver
): Promise<GeneratedPokemon> {
  const basicPokemonInfo = getRandomFromList(random, possiblePokemon);
  const info = await pokemonResolver(basicPokemonInfo.no);

  const allPossibleMovesList = await getPossibleMoves(
    info,
    basicMoveResolver,
    moveListResolver
  );

  const moves = await getRandomMoves(
    random,
    allPossibleMovesList,
    4,
    moveResolver
  );
  const nature = getRandomFromList(random, possibleNatures);
  const ability = getRandomFromList(random, possibleAbilities);

  const pokemonInfo: GeneratedPokemonNoExport = {
    info,
    nature,
    ability,
    moves,
  };

  const showdownExport = getShowdownExport(pokemonInfo);

  return {
    ...pokemonInfo,
    showdownExport,
  };
}
