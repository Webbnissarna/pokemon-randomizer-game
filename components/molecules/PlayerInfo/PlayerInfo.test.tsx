import { screen } from "@testing-library/react";
import { render } from "../../../.jest/render";
import PlayerInfo from ".";
import { GeneratedPokemon } from "../../../utils/Generator/generator.types";

describe("PlayerInfo", () => {
  it("renders", () => {
    const func = () =>
      render(
        <PlayerInfo
          playerName=""
          variant="br"
          ballCount={0}
          pickedPokemon={[]}
        />
      );

    expect(func).not.toThrow();
  });

  it("shows the player name", () => {
    const playerName = "Random player name";

    render(
      <PlayerInfo
        playerName={playerName}
        variant={"br"}
        ballCount={0}
        pickedPokemon={[]}
      />
    );

    const element = screen.getByText(playerName);

    expect(element).toBeTruthy();
  });

  it("shows the correct amount of balls", () => {
    const ballCount = 6;

    render(
      <PlayerInfo
        playerName=""
        variant={"br"}
        ballCount={ballCount}
        pickedPokemon={[]}
      />
    );

    const elements = screen.getAllByRole("button");

    expect(elements).toHaveLength(ballCount);
  });

  it("handles invalid ball count", () => {
    const ballCount = -1;

    render(
      <PlayerInfo
        playerName=""
        variant={"br"}
        ballCount={ballCount}
        pickedPokemon={[]}
      />
    );

    const element = screen.queryByRole("button");

    expect(element).toBeNull();
  });

  it("always shows enough ball for all picked PokÃ©mon", () => {
    const ballCount = 6;
    const pickedCount = 3;
    const mockPicked = Array(pickedCount).fill({}) as GeneratedPokemon[];

    render(
      <PlayerInfo
        playerName=""
        variant={"br"}
        ballCount={ballCount}
        pickedPokemon={mockPicked}
      />
    );

    const elements = screen.getAllByRole("button");
    const disabledElements = elements.filter((element) =>
      element.hasAttribute("disabled")
    );

    expect(disabledElements).toHaveLength(ballCount - pickedCount);
  });
});
