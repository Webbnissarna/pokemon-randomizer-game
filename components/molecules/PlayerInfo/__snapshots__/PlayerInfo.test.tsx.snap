// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayerInfo renders 1`] = `
<DocumentFragment>
  <script>
    (function() { try {
  var mode = localStorage.getItem('theme-ui-color-mode');
  if (!mode) return
  document.documentElement.classList.add('theme-ui-' + mode);
} catch (e) {} })();
  </script>
  <div
    class="css-qznt7n-Box"
  >
    <span
      class="css-2vsnqm-Box"
    />
    <div
      class="css-p1luab-Box"
    />
  </div>
</DocumentFragment>
`;
